numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -426,16 +440,25 @@ UNICODE_setitem(PyObject *op, void *ov, void *vap)
     char *buffer;
 #endif
 
-    if (!PyBytes_Check(op) && !PyUnicode_Check(op) &&
-            PySequence_Check(op) && PySequence_Size(op) > 0) {
+    /* Handle case of assigning from an array scalar */
+    if (PyArray_Check(op) && PyArray_NDIM((PyArrayObject *)op) == 0) {
```

Comment:
Switched to `PyArray_IsZeroDim`. The way this is currently written (recursing) doesn't lend itself particularly well to being placed in a function.

And recursing is actually a pretty tidy way to deal with the refcounting here - the alternative being replacing every early return with a `goto fail;`